src/mvp/MvpServer.cpp

```diff
@@ -341,6 +345,32 @@ void MvpServer::AttackShip(std::shared_ptr<Connection> client,
     ship.attackTargetID = targetID;
 }
 
+void MvpServer::CheckTreasureIsland(int treasure_x, int treasure_y,
+                                    PlayerData &player) {
+    std::map<uint16_t, admirals::mvp::ShipData> &shipList = player.ships;
+    for (int y = treasure_y - 1; y <= treasure_y + 1; y++) {
+        if (y < 0 || y >= BOARD_SIZE) {
+            continue;
+        }
+        for (int x = treasure_x - 1; x <= treasure_x + 1; x++) {
+            if (x < 0 || x >= BOARD_SIZE) {
+                continue;
+            }
+            const uint16_t shipId = m_board[x][y];
+            if (shipId != 0) {
+                if (auto it = shipList.find(shipId); it != shipList.end()) {
+                    player.coins += 5;
+                }
+            }
+        }
+    }
+}
```

Comment:
This is quite funny because exactly this problem was part of advent of code just a couple of days ago. I would rewrite this function in the following way:

```c++
static bool coordinateInBounds(int x, int y) {
    return x >= 0 && x < BOARD_SIZE && y >= 0 && y < BOARD_SIZE;
}

void MvpServer::CheckTreasureIsland(int tx, int ty, PlayerData &player) {
    for (int dx = -1; dx <= 1; dx++) {
        for (int dy = -1; dy <= 1; dy++) {
            const uint16_t shipId = m_board[tx + dx][ty + dy];
            if ((dx != 0 || dy != 0) && coordinateInBounds(tx + dx, ty + dy) &&
                shipId != 0 && player.ships.contains(shipId)) {
                player.coins += 100;
                return;
            }
        }
    }
}

```

Reply:
While we're at it I would also define some constant called `REWARD_TREASURE_ISLAND` that says how many coins should be awarded for being close to the treasure island.

Reply:
> This is quite funny because exactly this problem was part of advent of code just a couple of days ago. I would rewrite this function in the following way:
> 
> ```c++
> static bool coordinateInBounds(int x, int y) {
>     return x >= 0 && x < BOARD_SIZE && y >= 0 && y < BOARD_SIZE;
> }
> 
> void MvpServer::CheckTreasureIsland(int tx, int ty, PlayerData &player) {
>     for (int dx = -1; dx <= 1; dx++) {
>         for (int dy = -1; dy <= 1; dy++) {
>             const uint16_t shipId = m_board[tx + dx][ty + dy];
>             if ((dx != 0 || dy != 0) && coordinateInBounds(tx + dx, ty + dy) &&
>                 shipId != 0 && player.ships.contains(shipId)) {
>                 player.coins += 100;
>                 return;
>             }
>         }
>     }
> }
> ```

This could cause errors as  the board is indexed without checking the bounds. It also changes how the coins are collected, only allowing one ship from each player to collect coins at once. We should use `player.ships.contains(shipId)` instead of `auto it = shipList.find(shipId); it != shipList.end()` but otherwise the original solution will work better.